syntax = "proto3";

import "google/protobuf/empty.proto";

package task_grpc;

message User {
    uint64 id = 1;
    string login = 2;
}

message CreateUserRequest {
    string login = 1;
}

message CreateUserResponse {
    User user = 1;
}

message Task {
    uint64 id = 1;
    string creator_login = 2;
    string name = 3;
    string text = 4;
}

message CreateTaskRequest {
    string creator_login = 1;
    string name = 2;
    string text = 3;
}

message CreateTaskResponse {
    Task task = 1;
}

message UpdateTaskRequest {
    uint64 id = 1;
    string creator_login = 2;
    string new_name = 3;
    string text = 4;
}

message UpdateTaskResponse {
    Task task = 1;
}

message DeleteTaskRequest {
    string creator_login = 1;
    uint64 task_id = 2;
}

message DeleteTaskResponse {
    Task task = 1;
}

message GetTaskRequest {
    uint64 id = 1;
}

message GetTaskResponse {
    Task task = 1;
}

message ListTasksRequest {
    int32 page = 1;
    int32 limit = 2;
}

message ListTasksResponse {
    repeated Task tasks = 1;
}

service TaskManagerServer {
    rpc CreateUser (CreateUserRequest) returns (CreateUserResponse) {}
    rpc CreateTask (CreateTaskRequest) returns (CreateTaskResponse) {}
    rpc UpdateTask (UpdateTaskRequest) returns (UpdateTaskResponse) {}
    rpc DeleteTask (DeleteTaskRequest) returns (DeleteTaskResponse) {}
    rpc GetTask (GetTaskRequest) returns (GetTaskResponse) {}
    rpc ListTasks (ListTasksRequest) returns (ListTasksResponse) {}
}
